Support live playback of demos while they are being recorded

Robyt3 opened the pull request:
Check if more ticks are still being written to the demo 1 second after starting playback to detect a live demo. For live demos, each tick incrementally scan for new tickmarkers and keyframes after the last keyframe to properly update the total time while playing the demo. Detect if the demo stopped being recorded by checking if no more tickmarkers have been written within 2 seconds.

Prevent seeking closer than 2 seconds to the end of live demos to avoid demo playback errors due to incomplete reads of the last demo chunk data which may still be written to by a demo recorder. Also reset the demo player speed to the default when fast-forwarding to the end of a live demo to prevent playback errors.

Show a red/grey dot indicator next to the demo seekbar for live demos. The indicator is red to indicate playback being considered live when viewing the last 3 seconds of the live demo. Clicking the live indicator will seek to the most recent viewable time.

![image](https://github.com/user-attachments/assets/e8785cee-23e1-4815-a9d6-0655f5e47033)

![image](https://github.com/user-attachments/assets/3737cec6-55a7-4666-b773-afddbc36d6c5)

Furthermore, support playing truncated demos again when possible instead of failing to open them entirely. Keyframes will now only be read until the truncation occurs. Skipping to the end of a truncated demo may still end playback with an error due to the chunk data being incomplete. A truncated demo file is indistinguishable from a live demo being read in an incomplete state. The detection for live demos is deactivated after 15 seconds if it keeps failing, to avoid unnecessary scanning of files that are likely truncated and not live demos.

## Checklist

- [X] Tested the change ingame
- [X] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

Jupeyy commented:
Do you really want to keep the live option?
Like can you imagine anything useful with it?

How often do you open a second client just to view a second demo at the same time ^^

Rest looks fine tho

Robyt3 commented:
I moved the other commits to #10137. Changed to draft because this seems like a feature that we need to discuss if we want the complexity. I can imagine it being useful for server demos to watch live what is happening on your server without joining it. @jxsl13 seemed interested in the feature on Discord.

Live demos would also need some testing whether they work on other systems as well.

murpii commented:
> Do you really want to keep the live option? Like can you imagine anything useful with it?
> 
> How often do you open a second client just to view a second demo at the same time ^^
> 
> Rest looks fine tho

I do that quite regularly

Jupeyy commented:
> I do that quite regularly

Well I meant starting a second client to watch an actively written demo, not any demo ^^

But yeah as long as the live button is never shown for normal demos it's at least not confusing for the users.

ChillerDragon commented:
> How often do you open a second client just to view a second demo at the same time ^^

I do it often to see what happend earlier when I do not want to disconnect in game. But then I am usually not interested in anything that happend after I started the playback. I am often not sure if it somehow breaks the demo. So this being fully supported could give players like me a peace of mind even if they don't need the full live feature.

The live feature could be used to play on one client and stream a demo without hud on another client if the server only allows one client per ip.

Robyt3 commented on src/engine/shared/demo.cpp:
```diff
@@ -840,11 +868,12 @@ int CDemoPlayer::Load(class IStorage *pStorage, class IConsole *pConsole, const
 
 	// scan the file for interesting points
-	if(!ScanFile())
+	if(ScanFile(EScanFileMode::FULL) != EScanFileResult::SUCCESS)
```
Should probably also handle `EScanFileResult::ERROR_RECOVERABLE` here and in `ScanFile` with `EScanFileMode::FULL` without clearing all keyframes because the first scan might fail at the end if it happens at the same time as the next chunk is being written.

ChillerDragon commented:
Works on my machine

l-ouis commented:
Servers streaming live demos to a 'live' section ingame could be pretty cool i'd imagine

ChillerDragon commented:
Thanks
